src/sctools/barcode.py

```diff
@@ -112,3 +113,66 @@ def from_iterable_bytes(cls, iterable, barcode_length):
         """construct an ObservedBarcodeSet from an iterable of bytes barcodes"""
         tbe = TwoBit(barcode_length)
         return cls(Counter(tbe.encode(b) for b in iterable), barcode_length=barcode_length)
+
+
+class ErrorsToCorrectBarcodesMap:
+
+    def __init__(self, errors_to_barcodes):
+        """use a hash map of erroneous_barcode -> true barcode to correct errors in cell barcodes
+
+        :param dict errors_to_barcodes:
+        """
+        if not isinstance(errors_to_barcodes, Mapping):
+            raise TypeError('errors_to_barcodes must be a mapping of erroneous barcodes to correct '
+                            'barcodes, not %a' % type(errors_to_barcodes))
+        self._map = errors_to_barcodes
+
+    def correct_barcode(self, barcode):
+        return self._map[barcode]
+
+    @staticmethod
+    def _prepare_single_base_error_hash_table(iterable):
+        """
+
+        :param Iterable iterable: iterable of string barcodes
+        :return:
+        """
+        error_map = {}
+        for barcode in iterable:
+
+            # include correct barcode
+            error_map[barcode] = barcode
+
+            # include all single-base errors
+            for i, nucleotide in enumerate(barcode):
+                errors = set('ACGTN')
+                errors.discard(nucleotide)
+                for e in errors:
+                    error_map[barcode[:i] + e + barcode[i + 1:]] = barcode
+        return error_map
+
+    @classmethod
+    def single_hamming_errors_from_whitelist(cls, whitelist_file):
+        """factory method to generate instance of class from a file containing "correct" barcodes
+
+        :param str whitelist_file:
+        :return CorrectBarcodeErrorsHash:
```

Comment:
How is this different from `ErrorsToCorrectBarcodesMap`?

Reply:
Good catch; failed to refactor the docstrings during development. It should be as you say.